build_tools/android/common/benchmark_description.py

```diff
@@ -0,0 +1,280 @@
+                  capture_output=True,
+                  verbose=verbose).stdout.strip()
+  model = re.sub(r"\W+", "-", model)
+  return model
+
+
+def get_android_cpu_abi(verbose: bool = False) -> str:
+  """Returns the CPU ABI for the Android device."""
+  return execute(["adb", "shell", "getprop", "ro.product.cpu.abi"],
+                 capture_output=True,
+                 verbose=verbose).stdout.strip()
+
+
+def get_android_cpu_features(verbose: bool = False) -> Sequence[str]:
+  """Returns the CPU features for the Android device."""
+  cpuinfo = execute(["adb", "shell", "cat", "/proc/cpuinfo"],
+                    capture_output=True,
+                    verbose=verbose).stdout.strip()
+  features = []
+  for line in cpuinfo.splitlines():
+    if line.startswith("Features"):
+      _, features = line.split(":")
+      return features.strip().split()
+  return features
+
+
+def get_android_gpu_name(verbose: bool = False) -> str:
+  """Returns the GPU name for the Android device."""
+  vkjson = execute(["adb", "shell", "cmd", "gpu", "vkjson"],
+                   capture_output=True,
+                   verbose=verbose).stdout.strip()
+  vkjson = json.loads(vkjson)
+  name = vkjson["devices"][0]["properties"]["deviceName"]
+
+  # Perform some canonicalization:
+
+  # - Adreno GPUs have raw names like "Adreno (TM) 650".
+  name = name.replace("(TM)", "")
+
+  # Replace all consecutive non-word characters with a single hypen.
+  name = re.sub(r"\W+", "-", name)
+
+  return name
+
+
+@dataclass
+class AndroidDeviceInfo:
+  """An object describing the current Android Device.
+
+  It includes the following phone characteristics:
+  - model: the product model, e.g., 'Pixel-4'
+  - cpu_abi: the CPU ABI, e.g., 'arm64-v8a'
+  - cpu_features: the detailed CPU features, e.g., ['fphp', 'sve']
+  - gpu_name: the GPU name, e.g., 'Mali-G77'
+  """
+
+  model: str
+  cpu_abi: str
+  cpu_features: Sequence[str]
+  gpu_name: str
+
+  def __str__(self):
+    features = ", ".join(self.cpu_features)
+    params = [
+        f"model='{self.model}'",
+        f"cpu_abi='{self.cpu_abi}'",
+        f"gpu_name='{self.gpu_name}'",
+        f"cpu_features=[{features}]",
+    ]
+    params = ", ".join(params)
+    return f"Android device <{params}>"
+
+  def get_arm_arch_revision(self) -> str:
+    """Returns the ARM architecture revision."""
+    if self.cpu_abi != "arm64-v8a":
+      raise ValueError("Unrecognized ARM CPU ABI; need to update the list")
+
+    # CPU features for ARMv8 revisions.
+    # From https://en.wikichip.org/wiki/arm/armv8#ARMv8_Extensions_and_Processor_Features
+    rev1_features = ["atomics", "asimdrdm"]
+    rev2_features = [
+        "fphp", "dcpop", "sha3", "sm3", "sm4", "asimddp", "sha512", "sve"
+    ]
+
+    rev = "ARMv8-A"
+    if any([f in self.cpu_features for f in rev1_features]):
+      rev = "ARMv8.1-A"
+    if any([f in self.cpu_features for f in rev2_features]):
+      rev = "ARMv8.2-A"
+    return rev
```

Comment:
What about 8.3? Should we error out if there are features we don't recognize?

Reply:
In the Android world right now we are pretty stuck at v8.2. Copying my previous comment (https://github.com/google/iree/pull/5837#discussion_r633698271):

> It could be. But in Android land, pretty much all SoCs top at v8.2-A. See https://en.wikipedia.org/wiki/Comparison_of_ARMv8-A_cores (I don't think the Kryo 6XX having v8.4-A is correct there.) So we won't need to worry about it for now (and probably quite some time in the future). :)

Even if there is a v8.2+ chip, treating them as v8.2 shouldn't hurt because they should support v8.2 features.